piptools/utils.py

```diff
@@ -223,6 +239,55 @@ def get_hashes_from_ireq(ireq: InstallRequirement) -> Set[str]:
     return result
 
 
+def abs_ireq(ireq: InstallRequirement, from_dir: str) -> InstallRequirement:
+    """
+    Return the given InstallRequirement if its source isn't a relative path;
+    Otherwise, return a new one with the relative path rewritten as absolute.
+
+    In this case, an extra attribute is added: _was_relative,
+    which is always True when present at all.
+    """
+    if ireq.link is None or not ireq.link.is_file:
+        return ireq
+
+    start_dir = os.path.abspath(os.curdir)
+    try:
+        os.chdir(from_dir)
+        local_abspath = os.path.abspath(ireq.local_file_path)
+    finally:
+        os.chdir(start_dir)
```

Comment:
Looks like a job a for context namager:
```suggestion
    with chdir(from_dir):
        local_abspath = os.path.abspath(ireq.local_file_path)
```

Reply:
Does that `chdir` context manager already exist in the code or any libraries used?

As I said, I know `plumbum` path handling offers context managers OOTB, but I don't know if the various path tools used in this project do. Keep in mind that this is the *single* place in the code this logic is needed, and `try`+`finally` is accomplishing what we want fairly concisely.

If we were needing this a lot (twice?), or if we got this context manager "for free", or if `try`+`finally` doesn't do exactly what's needed, I'll agree. Otherwise it doesn't seem worth it. Maybe I'm missing a benefit.

But I don't really need to agree. If you or another maintainer still prefer we wrap this in a new custom context manager I'll write it.

Reply:
The main benefit is readability. When I see `try-finally` with cleaning something in `finally` block I always think about context managers because that's what they are for. Even if it's used only once it's good for decoupling implementation details.

Since it's just a matter of taste feel free to ignore the suggestion. No worries 👍🏼

Reply:
Actually I just realize a likely second need to switch folders safely: I plan to update `pip-sync` to result relative ireq paths in files as relative to the file, rather than the invocation directory.

I would still like to know, do any of these project's tools already provide the context manager, or need we reinvent it?

Also, should the planned `pip-sync` change be part of this PR, or separate?

Reply:
I did indeed come across a second case for this, in our `parse_requirements` wrapper, so I added a context manager. Thanks!